### ISSUE TITLE
Typed `this` in JavaScript Object Literals

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#16184]
From https://github.com/Microsoft/vscode/issues/27693

**TypeScript Version:**  2.3.4

```js
const obj = {
    prop: 2,

    method() {
        this.prop|
    }
}
```

https://github.com/Microsoft/TypeScript/pull/14141 added typing for `this` in object literals in certain cases. Should this also be applied in regular JavaScript for intellisense, perhaps in a less strict form? We currently do not return any suggestions for `this.|` in `obj`

**Comment by mhegazy** [comment]
referenced in https://github.com/Microsoft/TypeScript/issues/14617

**Comment by mhegazy** [comment]
we should enable the inference by default for .js files (similar to `--noImplicitAny` contingent inferences)

**Comment by n-peugnet** [comment]
ok, this is exactly what I wanted, so I think this indeed should also be applied in regular JavaScript.

**Comment by sandersn** [comment]
Fix is up at #17252
